hive/agents/ppo.py

```diff
@@ -0,0 +1,379 @@
+        self._actor_critic = PPOActorCriticNetwork(
+            network,
+            actor_net,
+            critic_net,
+            network_output_shape,
+            self._action_space,
+            self._continuous_action,
+        ).to(self._device)
+
+        self._actor_critic.apply(self._init_fn)
+
+    def train(self):
+        """Changes the agent to training mode."""
+        super().train()
+        self._actor_critic.train()
+
+    def eval(self):
+        """Changes the agent to evaluation mode."""
+        super().eval()
+        self._actor_critic.eval()
+
+    def preprocess_update_info(self, update_info):
+        """Preprocesses the :obj:`update_info` before it goes into the replay buffer.
+
+        Args:
+            update_info: Contains the information from the current timestep that the
+                agent should use to update itself.
+        """
+        if self._reward_clip is not None:
+            update_info["reward"] = np.clip(
+                update_info["reward"], -self._reward_clip, self._reward_clip
+            )
+        preprocessed_update_info = {
+            "observation": update_info["observation"],
+            "action": update_info["action"],
+            "reward": update_info["reward"],
+            "done": update_info["done"],
+            "logprob": self._logprob,
+            "values": self._value,
+            "returns": np.empty(self._value.shape),
+            "advantages": np.empty(self._value.shape),
+        }
+        if "agent_id" in update_info:
+            preprocessed_update_info["agent_id"] = int(update_info["agent_id"])
+
+        return preprocessed_update_info
+
+    def preprocess_update_batch(self, batch):
+        """Preprocess the batch sampled from the replay buffer.
+
+        Args:
+            batch: Batch sampled from the replay buffer for the current update.
+
+        Returns:
+                - Preprocessed batch.
+        """
+        for key in batch:
+            batch[key] = torch.tensor(batch[key], device=self._device)
+        return batch
+
+    @torch.no_grad()
+    def get_action_logprob_value(self, observation):
+        """Returns the action, logprob, and value for the agent
+
+        Args:
+            observation: The current observation.
+        """
+        observation = torch.tensor(
+            np.expand_dims(observation, axis=0), device=self._device
+        ).float()
+        action, logprob, _, value = self._actor_critic(observation)
+        action = action.cpu().detach().numpy()
+        logprob = logprob.cpu().numpy()
+        value = value.cpu().numpy()
+        action = action[0]
+
+        return action, logprob, value
+
+    @torch.no_grad()
+    def act(self, observation):
+        """Returns the action for the agent.
+
+        Args:
+            observation: The current observation.
+        """
+        action, logprob, value = self.get_action_logprob_value(observation)
+        self._logprob = logprob
+        self._value = value
```

Comment:
Could you put this into a single state dictionary? It will be easier to port when I add the changes for parallel environments, as the agents will become a bit more stateless (they won't store these state variables, instead they will pass those dictionaries around)